File: backend/fcm.py
#!/usr/bin/env python3

from pymongo import MongoClient
from firebase_admin import credentials
from firebase_admin import messaging
import firebase_admin
from datetime import * 
import os

# connect to mongodb 
client = MongoClient(os.getenv("MONGO_URL"))
db = client.citybreak

# authorize firebase project use
cred = credentials.Certificate(os.getenv("CB_FIREBASE_CREDENTIALS")) 
firebase_app = firebase_admin.initialize_app(cred)

def sendDailyNotification():
    cities = list(db.cities.find())

    for city in cities:
        events = list(db.events.find({"city_id": city['city_id']}).sort("date_time", 1))
        tomorrow_events = 0
            
        for event in events:
            today = datetime.now().date()
            diff = event['date_time'].date() - today
            
            if diff.days == 1:
                tomorrow_events += 1
            elif diff.days > 1:
                break

        if tomorrow_events == 0:
            continue

        title = "Mâine în orașul tău"
        body = ""

        if tomorrow_events == 1:
            body = "Un eveniment nou"
        elif tomorrow_events == 2:
            body = "Două evenimente noi"
        elif tomorrow_events < 20:
            body = str(tomorrow_events) + " evenimente noi"
        else:
            body = str(tomorrow_events) + " de evenimente noi" 
        
        message = messaging.Message(
            notification = messaging.Notification(
                title=title,
                body=body,
            ),
            data = {
                'type': 'tomorrow_events',
            },
            android = messaging.AndroidConfig(
                ttl = timedelta(days=2),
            ),
            apns = messaging.APNSConfig(
                headers = {
                    "apns-expiration": "172800",
                }
            ),
            topic = city['city_id'],
        )

        messaging.send(message)

def sendNewEventNotification(name, _id, city_id, city_name):
    message = messaging.Message(
        notification = messaging.Notification(
            title="Un nou eveniment in " + city_name + "!",
            body=name,
        ),
        data = {
            'type': 'event',
            'id': str(_id),
        },
        topic = city_id,
    )

    messaging.send(message)

if __name__ == '__main__':
    sendDailyNotification()

File: api/src/utils/fcm.ts
import admin, { messaging } from "firebase-admin";
import { MongoClient } from "mongodb";
import { City } from "../models/cityModel";
import { Event } from "../models/eventModel";
import dotenv from "dotenv";
import path from "path";
import dayjs from "dayjs";
import isTomorrow from "dayjs/plugin/isTomorrow";

dotenv.config({ path: path.join(__dirname, "..", "..", ".env") });

const FIREBASE_CREDENTIALS = process.env.CB_FIREBASE_CREDENTIALS || "";
const MONGO_URL = process.env.CB_MONGODB_URL || "";

admin.initializeApp({
  credential: admin.credential.cert(FIREBASE_CREDENTIALS),
});

dayjs.extend(isTomorrow);

const sendDailyNotification = async () => {
  const client = new MongoClient(MONGO_URL);
  const db = client.db("citybreak");
  const citiesCollection = db.collection<City>("cities");
  const eventsCollection = db.collection<Event>("events");

  const cities = await citiesCollection.find().toArray();

  for (let city of cities) {
    const events = await eventsCollection
      .find({ city_id: city.city_id })
      .sort({ date_time: 1 })
      .toArray();
    let tomorrow_events = 0;

    for (let event of events) {
      const event_date = dayjs(event.date_time);

      if (event_date.isTomorrow()) {
        tomorrow_events += 1;
      } else if (event_date.isAfter(dayjs().add(1, "days"))) {
        break;
      }
    }

    if (tomorrow_events === 0) {
      continue;
    }

    const title = "Mâine în orașul tău";
    let body = "";

    if (tomorrow_events === 1) {
      body = "Un eveniment nou";
    } else if (tomorrow_events === 2) {
      body = "Două evenimente noi";
    } else if (tomorrow_events < 20) {
      body = `${tomorrow_events} evenimente noi`;
    } else {
      body = `${tomorrow_events} de evenimente noi`;
    }

    const H18inMS: number = 64800000; // 18 hours in ms
    const appleTTL = dayjs().add(18, "hours").unix();

    const message: messaging.Message = {
      notification: {
        title: title,
        body: body,
      },
      data: {
        type: "tomorrow_events",
      },
      android: {
        // priority: "high" // wake devices from doze mode (bad for energy)
        ttl: H18inMS, // live notification for 18 hours
      },
      apns: {
        headers: {
          "apns-expiration": appleTTL.toString(), // 18 hours from now
          // "apns-priority": "10" // send notification immediately (bad for energy, default: 5)
        },
      },
      topic: city.city_id,
    };

    admin.messaging().send(message);
  }
};

export const sendNewEventNotification = (
  name: string,
  _id: string,
  city_id: string,
  city_name: string,
) => {
  const H18inMS: number = 64800000; // 18 hours in ms
  const appleTTL: number = Date.now() + H18inMS; // apns-expiration should be in UNIX epoch time format

  const message = {
    notification: {
      title: "Un nou eveniment in " + city_name + "!",
      body: name,
    },
    data: {
      type: "event",
      id: _id,
    },
    android: {
      // priority: "high" // wake devices from doze mode (bad for energy)
      ttl: H18inMS, // live notification for 18 hours
    },
    apns: {
      headers: {
        "apns-expiration": appleTTL.toString(), // 18 hours from now
        // "apns-priority": "10" // send notification immediately (bad for energy, default: 5)
      },
    },
    topic: city_id,
  };

  admin.messaging().send(message);
};
